passes/sat/qbfsat.cc

```diff
@@ -0,0 +1,522 @@
+		std::vector<RTLIL::SigBit> value_bv;
+		value_bv.reserve(wire->width);
+		for (char c : hole_value)
+			value_bv.push_back(c == '1'? RTLIL::S1 : RTLIL::S0);
+		std::reverse(value_bv.begin(), value_bv.end());
+		module->connect(wire, value_bv);
+	}
+}
+
+void allconstify_inputs(RTLIL::Module *module, const std::set<std::string> &input_wires) {
+	for (auto &n : input_wires) {
+		RTLIL::Wire *input = module->wire(n);
+#ifndef NDEBUG
+		log_assert(input != nullptr);
+#endif
+
+		RTLIL::Cell *allconst = module->addCell("$allconst$" + n, "$allconst");
+		allconst->setParam(ID(WIDTH), input->width);
+		allconst->setPort(ID::Y, input);
+		allconst->set_src_attribute(input->get_src_attribute());
+		input->port_input = false;
+		log("Replaced input %s with $allconst cell.\n", n.c_str());
+	}
+	module->fixup_ports();
+}
+
+void assume_miter_outputs(RTLIL::Module *module) {
+	std::vector<RTLIL::Wire *> wires_to_assume;
+	for (auto w : module->wires())
+		if (w->port_output) {
+			if (w->width == 1)
+				wires_to_assume.push_back(w);
+		}
+	if (wires_to_assume.size() == 0)
+		return;
+	else {
+		log("Adding $assume cell for outputs: ");
+		for (auto w : wires_to_assume)
+			log("\"%s\" ", w->name.c_str());
+		log("\n");
+	}
+
+	std::vector<RTLIL::Wire *>::size_type i;
+	while (wires_to_assume.size() > 1) {
+		std::vector<RTLIL::Wire *> buf;
+		for (i = 0; i + 1 < wires_to_assume.size(); i += 2) {
+			std::stringstream strstr; strstr << i;
+			RTLIL::Wire *and_wire = module->addWire("\\_qbfsat_and_" + strstr.str(), 1);
+			module->addLogicAnd("$_qbfsat_and_" + strstr.str(), wires_to_assume[i], wires_to_assume[i+1], and_wire, false, wires_to_assume[i]->get_src_attribute());
+			buf.push_back(and_wire);
+		}
+		wires_to_assume.swap(buf);
+	}
+
+#ifndef NDEBUG
+	log_assert(wires_to_assume.size() == 1);
+#endif
+	module->addAssume("$assume_qbfsat_miter_outputs", wires_to_assume[0], RTLIL::S1);
+}
+
+QbfSolutionType qbf_solve(RTLIL::Module *mod, const QbfSolveOptions &opt) {
+	QbfSolutionType ret;
+	std::string yosys_smtbmc_exe = proc_self_dirname() + "yosys-smtbmc";
+	std::string smtbmc_warning = "z3: WARNING:";
+
+	std::string tempdir_name = "/tmp/yosys-z3-XXXXXX";
+	tempdir_name = make_temp_dir(tempdir_name);
+	std::string smt2_command = "write_smt2 -stbv -wires " + tempdir_name + "/problem.smt2";
+#ifndef NDEBUG
+	log_assert(mod->design != nullptr);
+#endif
+	Pass::call(mod->design, smt2_command);
+
+	//Execute and capture stdout from `yosys-smtbmc -s z3 -t 1 -g --binary [--dump-smt2 <file>]`
+	{
+		fflush(stdout);
+		bool keep_reading = true;
+		int status = 0;
+		int retval = 0;
+		char buf[1024] = {0};
+		std::string linebuf = "";
+		std::string cmd = yosys_smtbmc_exe + " -s z3 -t 1 -g --binary " + (opt.dump_final_smt2? "--dump-smt2 " + opt.dump_final_smt2_file + " " : "") + tempdir_name + "/problem.smt2 2>&1";
+		log("Launching \"%s\".\n", cmd.c_str());
+
+#ifndef EMSCRIPTEN
```

Comment:
Yes, it would be silently broken.  What is the correct way to disable the whole pass when `EMSCRIPTEN` is defined?

> Also, please have a look at the internal run_command API for a cross-platform substitute for popen.

Roger that.